CRM/Activity/BAO/Activity.php

```diff
@@ -341,126 +341,78 @@ public static function create(&$params) {
     }
 
     $activityId = $activity->id;
-    $sourceID = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Source');
-    $assigneeID = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Assignees');
-    $targetID = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Targets');
-
     $activityRecordTypes = [
-      'source_contact_id' => 'Activity Source',
-      'assignee_contact_id' => 'Activity Assignees',
-      'target_contact_id' => 'Activity Targets',
+      'source_contact_id' => CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Source'),
+      'assignee_contact_id' => CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Assignees'),
+      'target_contact_id' => CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_ActivityContact', 'record_type_id', 'Activity Targets'),
     ];
-    $activityContacts = array_intersect_key($params, $activityRecordTypes);
 
-    // Unset any which are a single id - this is just because they have traditionally been updated as
-    // opposed to delete + possible create for the others.
-    // Potentially we should nuance it for the others too....
-    foreach ($activityContacts as $index => $value) {
-      if ($value && !is_array($value)) {
-        unset($activityContacts[$index]);
+    $activityContacts = [];
+    // Cast to an array if we just have an integer. Index by record type id.
+    foreach ($activityRecordTypes as $key => $recordTypeID) {
+      if (isset($params[$key])) {
+        if (empty($params[$key])) {
+          $activityContacts[$recordTypeID] = [];
+        }
+        else {
+          foreach ((array) $params[$key] as $contactID) {
+            $activityContacts[$recordTypeID][$contactID] = $contactID;
+          }
+        }
       }
     }
 
     if ($action === 'edit' && !empty($activityContacts)) {
-      $wheres = [['activity_id', '=', $params['id']], ['record_type_id:name', 'IN', array_intersect_key($activityRecordTypes, $activityContacts)]];
+      $wheres = [];
+      foreach ($activityContacts as $recordTypeID => $contactIDs) {
+        if (!empty($contactIDs)) {
+          $wheres[$key] = "(record_type_id = $recordTypeID AND contact_id IN (" . implode(',', $contactIDs) . '))';
```

Comment:
The contactIDs might come from untrusted input to the function. Should probably db_escape these. e.g. array_map(array('CRM_Core_DAO', 'escapeString'), $contactIDs). Or maybe better CRM_Utils_Type::escapeAll() which doesn't look like it actually escapes but would error if not ints.

Reply:
we are already iterating through so I added a cast

<img width="712" alt="Screen Shot 2020-10-01 at 7 03 28 PM" src="https://user-images.githubusercontent.com/336308/94773713-d7222f00-0418-11eb-9de5-66e863b60d5e.png">